src/backend/distributed/master/master_stage_protocol.c

```diff
@@ -595,14 +602,109 @@ CreateShardsOnWorkers(Oid distributedRelationId, List *shardPlacements,
 
 
 /*
- * WorkerCreateShard applies DDL commands for the given shardId to create the
- * shard on the worker node. Commands are sent to the worker node over the
- * given connection.
+ * RelationShardListForShardCreate gets a shard interval and returns the placement
+ * accesses that would happen when a placement of the shard interval is created.
  */
-void
-WorkerCreateShard(Oid relationId, int shardIndex, uint64 shardId, List *ddlCommandList,
-				  List *foreignConstraintCommandList, MultiConnection *connection)
+static List *
+RelationShardListForShardCreate(ShardInterval *shardInterval)
 {
+	List *relationShardList = NIL;
+	RelationShard *relationShard = NULL;
+	Oid relationId = shardInterval->relationId;
+	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(relationId);
+	List *referencedRelationList = cacheEntry->referencedRelationsViaForeignKey;
+	List *referencingRelationList = cacheEntry->referencingRelationsViaForeignKey;
+	List *allForeignKeyRelations =
+		list_concat_unique_oid(referencedRelationList, referencingRelationList);
+	int shardIndex = -1;
+
+	/* record the placement access of the shard itself */
+	relationShard = CitusMakeNode(RelationShard);
+	relationShard->relationId = relationId;
+	relationShard->shardId = shardInterval->shardId;
+	relationShardList = list_make1(relationShard);
+
+	if (cacheEntry->partitionMethod == DISTRIBUTE_BY_HASH &&
+		cacheEntry->colocationId != INVALID_COLOCATION_ID)
+	{
+		shardIndex = ShardIndex(shardInterval);
+	}
+
+	/* all foregin constraint relations */
+	if (allForeignKeyRelations != NIL)
```

Comment:
Minor: We could also remove this `if` and unconditionally do the `foreach`. If you agree, then rename `relationIdCell` to `fkeyRelationIdCell`.